src/Microsoft.ML.TimeSeries/PredictionEngine.cs

```diff
@@ -0,0 +1,253 @@
+
+        internal PredictionEngine(IHostEnvironment env, ITransformer transformer, bool ignoreMissingColumns,
+            SchemaDefinition inputSchemaDefinition = null, SchemaDefinition outputSchemaDefinition = null)
+            : this(env, TransformerChecker(env, transformer), ignoreMissingColumns, inputSchemaDefinition, outputSchemaDefinition)
+        {
+        }
+
+        private static Func<Schema, IRowToRowMapper> TransformerChecker(IExceptionContext ectx, ITransformer transformer)
+        {
+            ectx.CheckValue(transformer, nameof(transformer));
+            ectx.CheckParam(transformer.IsRowToRowMapper, nameof(transformer), "Must be a row to row mapper");
+            return transformer.GetRowToRowMapper;
+        }
+
+        public void GetStatefulRows(IRow input, IRowToRowMapper mapper, Func<int, bool> active,
+            List<IStatefulRow> rows, out Action disposer, out IRow outRow)
+        {
+            Contracts.CheckValue(input, nameof(input));
+            Contracts.CheckValue(active, nameof(active));
+
+            disposer = null;
+            IRowToRowMapper[] innerMappers = new IRowToRowMapper[0];
+            if (mapper is CompositeRowToRowMapper)
+                innerMappers = ((CompositeRowToRowMapper)mapper).InnerMappers;
+
+            if (innerMappers.Length == 0)
+            {
+                bool differentActive = false;
+                for (int c = 0; c < input.Schema.ColumnCount; ++c)
+                {
+                    bool wantsActive = active(c);
+                    bool isActive = input.IsColumnActive(c);
+                    differentActive |= wantsActive != isActive;
+
+                    if (wantsActive && !isActive)
+                        throw Contracts.ExceptParam(nameof(input), $"Mapper required column '{input.Schema.GetColumnName(c)}' active but it was not.");
+                }
+
+                outRow = input;
+            }
+
+            // For each of the inner mappers, we will be calling their GetRow method, but to do so we need to know
+            // what we need from them. The last one will just have the input, but the rest will need to be
+            // computed based on the dependencies of the next one in the chain.
+            var deps = new Func<int, bool>[innerMappers.Length];
+            deps[deps.Length - 1] = active;
+            for (int i = deps.Length - 1; i >= 1; --i)
+                deps[i - 1] = innerMappers[i].GetDependencies(deps[i]);
+
+            IRow result = input;
+            for (int i = 0; i < innerMappers.Length; ++i)
+            {
+                Action localDisp;
+                if (innerMappers[i] is CompositeRowToRowMapper)
+                    GetStatefulRows(result, innerMappers[i], deps[i], rows, out localDisp, out result);
+                else
+                    result = innerMappers[i].GetRow(result, deps[i], out localDisp);
+
+                if (result is IStatefulRow)
+                    rows.Add((IStatefulRow)input);
```

Comment:
>input [](start = 43, length = 5)

`result`, not input.

I think this recursive function is not currently tested, and I think there's more bugs here. Please write a test that exercises this code

Reply:
Yep, I realized that when I added another test an hour back but thanks for also looking at this. In the next iteration this will be fixed.